File: Rust/Easy/0543_diameter_of_binary_tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::cmp;
use std::rc::Rc;
impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(root: &Option<Rc<RefCell<TreeNode>>>) -> (i32, i32) {
            match root {
                Some(r) => {
                    let (ld, lh) = dfs(&r.borrow().left);
                    let (rd, rh) = dfs(&r.borrow().right);
                    (ld.max(rd).max(lh + rh), cmp::max(lh, rh) + 1)
                }
                None => (0, 0),
            }
        }
        dfs(&root).0
    }
}

File: python/easy/0543_diameter_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(root) -> (int, int):
            if root is None:
                return (0, 0)
            ld, lh = dfs(root.left)
            rd, rh = dfs(root.right)
            return (max(ld, rd, lh + rh), max(lh, rh) + 1)
        md, mh = dfs(root)
        return md
